pkg/resource/deploy/source_query.go

```diff
@@ -0,0 +1,272 @@
+				Program:        src.runinfo.Program,
+				Args:           src.runinfo.Args,
+				Config:         config,
+				DryRun:         true,
+				Parallel:       math.MaxInt32,
+			})
+
+			// Check if we were asked to Bail.  This a special random constant used for that
+			// purpose.
+			if err == nil && bail {
+				return result.Bail()
+			}
+
+			if err == nil && progerr != "" {
+				// If the program had an unhandled error; propagate it to the caller.
+				err = errors.Errorf("an unhandled error occurred: %v", progerr)
+			}
+			return result.WrapIfNonNil(err)
+		}
+
+		// Communicate the error, if it exists, or nil if the program exited cleanly.
+		src.finChan <- run()
+	}()
+}
+
+// newQueryResourceMonitor creates a new resource monitor RPC server intended to be used in Pulumi's
+// "query mode".
+func newQueryResourceMonitor(builtins *builtinProvider) (*queryResmon, error) {
+
+	// Create our cancellation channel.
+	cancel := make(chan bool)
+
+	// New up an engine RPC server.
+	queryResmon := &queryResmon{
+		builtins: builtins,
+		cancel:   cancel,
+	}
+
+	// Fire up a gRPC server and start listening for incomings.
+	port, done, err := rpcutil.Serve(0, queryResmon.cancel, []func(*grpc.Server) error{
+		func(srv *grpc.Server) error {
+			pulumirpc.RegisterResourceMonitorServer(srv, queryResmon)
+			return nil
+		},
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	queryResmon.addr = fmt.Sprintf("127.0.0.1:%d", port)
+	queryResmon.done = done
+
+	return queryResmon, nil
+}
+
+// queryResmon is a pulumirpc.ResourceMonitor that is meant to run in Pulumi's "query mode". It
+// performs two critical functions:
+//
+// 1. Disallows all resource operations. `queryResmon` intercepts all resource operations and
+//    returns an error instead of allowing them to proceed.
+// 2. Services requests for stack snapshots. This is primarily to allow us to allow queries across
+//    stack snapshots.
+type queryResmon struct {
+	builtins *builtinProvider // provides builtins such as `getStack`.
+	addr     string           // the address the host is listening on.
+	cancel   chan bool        // a channel that can cancel the server.
+	done     chan error       // a channel that resolves when the server completes.
+}
+
+// Address returns the address at which the monitor's RPC server may be reached.
+func (rm *queryResmon) Address() string {
+	return rm.addr
+}
+
+// Cancel signals that the engine should be terminated, awaits its termination, and returns any
+// errors that result.
+func (rm *queryResmon) Cancel() error {
+	close(rm.cancel)
+	return <-rm.done
+}
+
+// Invoke performs an invocation of a member located in a resource provider.
+func (rm *queryResmon) Invoke(ctx context.Context, req *pulumirpc.InvokeRequest) (*pulumirpc.InvokeResponse, error) {
+	tok := tokens.ModuleMember(req.GetTok())
+	label := fmt.Sprintf("QueryResourceMonitor.Invoke(%s)", tok)
+
+	// Dispatch resource output request to default provider.
+	if tok == readStackResourceOutputs {
```

Comment:
It might be a little easier to read this function if we invert this `if` condition and immediately return if we're not doing the one allowed invoke. That way it's a lot easier to be absolutely sure that we're not accidentally following through the branch without returning and accidentally returning an error.